test/e2e/upgrade_test.go

```diff
@@ -0,0 +1,407 @@
+		}, e2eConfig.GetIntervals("default", "wait-available")...).Should(Succeed())
+		err = checkCertManagerAPI(upgradeClusterProxy)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Installing BMO on the upgrade cluster")
+		manifest, err := buildKustomizeManifest(bmoUpgradeFromKustomization)
+		Expect(err).NotTo(HaveOccurred())
+		err = upgradeClusterProxy.Apply(ctx, manifest)
+		Expect(err).NotTo(HaveOccurred())
+
+		bmoDeployment := &v1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "baremetal-operator-controller-manager",
+				Namespace: "baremetal-operator-system",
+			},
+		}
+		// Wait for it to become available
+		framework.WaitForDeploymentsAvailable(ctx, framework.WaitForDeploymentsAvailableInput{
+			Getter:     upgradeClusterProxy.GetClient(),
+			Deployment: bmoDeployment,
+		}, e2eConfig.GetIntervals("default", "wait-deployment")...)
+		// Set up log watcher
+		framework.WatchDeploymentLogsByName(ctx, framework.WatchDeploymentLogsByNameInput{
+			GetLister:  upgradeClusterProxy.GetClient(),
+			Cache:      upgradeClusterProxy.GetCache(ctx),
+			ClientSet:  upgradeClusterProxy.GetClientSet(),
+			Deployment: bmoDeployment,
+			LogPath:    filepath.Join(artifactFolder, "logs", bmoDeployment.GetNamespace()),
+		})
+
+		if e2eConfig.GetVariable("DEPLOY_IRONIC") != "false" {
+			// Install Ironic
+			By("Installing Ironic on the upgrade cluster")
+			ironicKustomization := e2eConfig.GetVariable("IRONIC_KUSTOMIZATION")
+			ironicManifest, err := buildKustomizeManifest(ironicKustomization)
+			Expect(err).NotTo(HaveOccurred())
+			err = upgradeClusterProxy.Apply(ctx, ironicManifest)
+			Expect(err).NotTo(HaveOccurred())
+
+			ironicDeployment := &v1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "baremetal-operator-ironic",
+					Namespace: "baremetal-operator-system",
+				},
+			}
+			// Wait for it to become available
+			framework.WaitForDeploymentsAvailable(ctx, framework.WaitForDeploymentsAvailableInput{
+				Getter:     upgradeClusterProxy.GetClient(),
+				Deployment: ironicDeployment,
+			}, e2eConfig.GetIntervals("ironic", "wait-deployment")...)
+			// Set up log watcher
+			framework.WatchDeploymentLogsByName(ctx, framework.WatchDeploymentLogsByNameInput{
+				GetLister:  upgradeClusterProxy.GetClient(),
+				Cache:      upgradeClusterProxy.GetCache(ctx),
+				ClientSet:  upgradeClusterProxy.GetClientSet(),
+				Deployment: ironicDeployment,
+				LogPath:    filepath.Join(artifactFolder, "logs", ironicDeployment.GetNamespace()),
+			})
+			namespace, cancelWatches = framework.CreateNamespaceAndWatchEvents(ctx, framework.CreateNamespaceAndWatchEventsInput{
+				Creator:   upgradeClusterProxy.GetClient(),
+				ClientSet: upgradeClusterProxy.GetClientSet(),
+				Name:      fmt.Sprintf("%s-%s", specName, util.RandomString(6)),
+				LogFolder: artifactFolder,
+			})
+		}
+	})
+
+	It("Should upgrade BMO to latest version", func() {
+		if e2eConfig.GetVariable("DEPLOY_IRONIC") != "false" {
```

Comment:
Why do we skip this if not deploying Ironic? This makes no sense to me

Reply:
Hmm. Why would we want it if we wouldn't provision anyway?

Reply:
Deploying Ironic has nothing to do with if we provision or not. It is possible for the user to USE_EXISTING_CLUSTER=true + DEPLOY_IRONIC=false so that they can set up their own cluster and Ironic and then run the tests against it.
Then we have the fixture tests, which is running the tests without Ironic completely, since BMO runs in test-mode. Then we don't deploy Ironic but we still go through all tests and we provision the BMHs. It will of course not do anything "real", but it is still a test that BMO behaves correctly and makes changes to the BMHs.

Reply:
What you said is true in general, but in this case isn't the point to check if the BMO can be upgraded from an older version? For the fixture test, we wouldn't be able to check if we can provision the BMH after upgrade anyway, so imo adding the BMH in the first place is a bit overkill, we should handle that in other tests. (I wouldn't mind adding it though if you think we should do so)

I think `USE_EXISTING_CLUSTER=true + DEPLOY_IRONIC=false` should not be allowed. That would make things too complicated. There are certain tasks (like provisioning) that wouldn't work without an Ironic, and we'd need separate configs for: "I don't have an ironic, but I don't need to install one, just ignore tests that requires an ironic", and "I have ironic already, don't attempt to install, just perform all the tests".

Reply:
There is no difference between fixture and ironic for the tests. We just run the same tests with BMO in test mode or BMO + Ironic. BMO in test mode can provision BMHs and that is what we test. Of course it is just fake when BMO is in test mode but that is beside the point.

> I think USE_EXISTING_CLUSTER=true + DEPLOY_IRONIC=false should not be allowed.

Why? In the tests you do not need to think about how Ironic is deployed. Just assume that it is there and working. If a user specifies `DEPLOY_IRONIC=false` it is up to them to make sure it works. They could be running Ironic outside the cluster, or we could be running with BMO in test mode, which does not require Ironic to begin with.

> There are certain tasks (like provisioning) that wouldn't work without an Ironic

All tasks will require Ironic in the real world, when actually interacting with baremetal servers. That does not mean that we must always deploy Ironic, or that Ironic is needed in all cases. We can run BMO in test mode for example, or Ironic can be set up manually outside the cluster.

Reply:
Got it. I forgot that we have the bmo test mode in fixture. Anw, my main point was that since this test was to check if BMO could be upgraded, we don't necessarily need to add all details in the fixture test (which could run for a while), but I'm not against running everything either.

Reply:
It is much harder to make special cases. Just run all tests but with different config for fixture. It does not hurt to test BMO upgrades in the fixture tests, in fact it could be very useful! We get a very quick test for checking if changes to the CRDs work and if BMO's webhook can handle conversions between new/old CRD versions.